grails-app/services/org/pih/warehouse/product/ProductIdentifierService.groovy

```diff
@@ -0,0 +1,201 @@
+            return generateSequentialProductIdentifier(productType)
+        }
+
+        return generateRandomProductIdentifier(productType)
+    }
+
+    private def generateSequentialProductIdentifier(ProductType productType) {
+        // If the product type does not have a custom identifier or does not contain character "0" which indicates the sequential part,
+        // then generate sequential product code from product type code
+        if (!productType.productIdentifierFormat || !productType.productIdentifierFormat.contains("0")) {
+            return generateSequentialProductIdentifierFromCode(productType)
+        }
+
+        def sequenceFormat = extractSequenceFormat(productType.productIdentifierFormat, Constants.DEFAULT_SEQUENCE_NUMBER_FORMAT_CHAR, 1)
+        def sequenceNumber = productTypeService.getAndSetNextSequenceNumber(productType)
+
+        return generateSequentialIdentifier(productType.productIdentifierFormat, sequenceFormat, sequenceNumber.toString())
+    }
+
+    private String generateSequentialProductIdentifierFromCode(ProductType productType) {
+        if (!productType) {
+            throw new IllegalArgumentException("Missing product type")
+        }
+
+        def defaultProductTypeId = configService.getProperty('openboxes.identifier.defaultProductType.id')
+        if (!productType.code && productType.id != defaultProductTypeId) {
+            throw new IllegalArgumentException("Can only generate sequential product code without specified product type code " +
+                    "for default product type with id: ${defaultProductTypeId}")
+        }
+
+        Integer sequenceNumber = productTypeService.getAndSetNextSequenceNumber(productType)
+        String sequenceNumberStr = generateSequenceNumber(sequenceNumber.toString())
+
+        return generate(IdentifierGeneratorParams.builder()
+                        .customKeys([
+                                "sequenceNumber": sequenceNumberStr,
+                                "code": productType.code,
+                        ])
+                        .build())
+    }
+
+    /**
+     * Generate random product identifier for specific productIdentifierFormat if it is set on the given product type,
+     * or for configured config.openboxes.identifier.product.format
+     */
+    private String generateRandomProductIdentifier(ProductType productType) {
+        // If the product type is null or the product type does not have a custom identifier,
+        // then generate product code from DEFAULT_PRODUCT_NUMBER_FORMAT
+        if (!productType || !productType.productIdentifierFormat) {
+            return generate()
+        }
+
+        // if the product type does not contain sequential part, then generate identifier basing on custom format
+        return generate(IdentifierGeneratorParams.builder()
+                .formatOverride(productType.productIdentifierFormat)
+                .build())
+    }
```

Comment:
yeah I tried to change as little of this class as possible when I brought it in so all that flow copy-pasted as is. I eventually want to refactor all of this so that it is muuuch simpler to use, but I'm saving that for future work

Reply:
Sounds good.